features/filesystem/storagelite/source/StorageLite.cpp

```diff
@@ -0,0 +1,1846 @@
+        entry = &ram_table[ind];
+        uint32_t curr_offset = entry->offset_and_flags & ram_offset_mask;
+        uint32_t ram_flags = entry->offset_and_flags & ram_flags_mask;
+        if (!curr_offset)  {
+            continue;
+        }
+        if (non_backup_only && (ram_flags & ram_has_backup_flag)) {
+            continue;
+        }
+        ret = read_record(_active_area, curr_offset, 0, 0, actual_name_size,
+                          0, 0, actual_data_size,
+                          false, false, false, false, false, hash, flags, nvstore_key,
+                          next_offset);
+        // Ignore rollback protect errors
+        if ((ret != STORAGELITE_SUCCESS) && (ret != STORAGELITE_RB_PROTECT_ERROR)) {
+            return ret;
+        }
+        if (!(flags & rollback_protect_flag)) {
+            continue;
+        }
+        os_ret = nvstore.remove(nvstore_key);
+        if ((os_ret != NVSTORE_SUCCESS) && (os_ret != NVSTORE_NOT_FOUND)) {
+            return STORAGELITE_OS_ERROR;
+        }
+    }
+    return ret;
+}
+
+
+int StorageLite::init(BlockDevice *bd, size_t max_files)
+{
+    ram_table_entry_t *ram_table;
+    area_state_e area_state[STORAGELITE_NUM_AREAS];
+    uint32_t init_attempts_val;
+    uint32_t next_offset;
+    uint32_t flags, hash;
+    uint32_t actual_data_size;
+    int os_ret;
+    int ret = STORAGELITE_SUCCESS;
+    uint16_t versions[STORAGELITE_NUM_AREAS];
+    uint16_t actual_name_size;
+    uint16_t nvstore_key;
+
+    if (_init_done) {
+        // Null BD signals the use of current one
+        if (!bd || (bd == _user_bd)) {
+            return STORAGELITE_SUCCESS;
+        }
+        // Reinitializing with a different BD - need to deinit first
+        deinit();
+    }
+    MBED_ASSERT(bd);
+
+    // This handles the case that init function is called by more than one thread concurrently.
+    // Only the one who gets the value of 1 in _init_attempts_val will proceed, while others will
+    // wait until init is finished.
+    init_attempts_val = core_util_atomic_incr_u32(&_init_attempts, 1);
+    if (init_attempts_val != 1) {
+        while (!_init_done) {
+            wait_ms(1);
+        }
+        return STORAGELITE_SUCCESS;
+    }
+
+    _max_files = max_files;
+    ram_table = new ram_table_entry_t[_max_files];
+    _ram_table = ram_table;
+    _num_ram_table_entries = 0;
+
+    NVStore& nvstore = NVStore::get_instance();
+    os_ret = nvstore.init();
+    MBED_ASSERT(!os_ret);
+
+    uint16_t curr_max_keys = nvstore.get_max_keys();
+
+    if ((size_t)(curr_max_keys - NVSTORE_NUM_PREDEFINED_KEYS) < max_files) {
+        // Need to have sufficient number of NVStore keys for rollback protection
+        // FIXME: Is this the correct way to do that?
+        nvstore.set_max_keys(max_files + NVSTORE_NUM_PREDEFINED_KEYS);
+        os_ret = nvstore.init();
+        MBED_ASSERT(!os_ret);
+    }
+
+    if (!_mutex) {
+        _mutex = new PlatformMutex;
```

Comment:
When it comes to "non trivial" member variables, I believe they should be dynamically allocated as part of the init phase (delaying the memory consumption as much as possible).

Reply:
Interesting, I view the mutex as a "trivial" member variable. The mutex doesn't have a buffer or anything, on some platforms it could be as small as an int.

Reply:
There was initially another reason, which eventually didn't apply here (was relevant in NVStore): Few of the tests used thread killing. Now mutexes (when taken) don't like this and sometimes get stuck when it happens. Deleting the mutex was a good workaround for this problem.